### PR TITLE
new shard on schema change

### PR DESCRIPTION
**Pull request by jshencode**
1. handle dimension table bootstrap
2. handle table addition event
3. clean up logging and bug fix

### CODE REVIEW

File: datanode/datanode.go [hunk 1]
```diff
@@ -479,25 +553,49 @@ func (d *dataNode) AssignShardSet(shardSet shard.ShardSet) {
...
+			d.memStore.AddTableShard(table, int(shard.ID()), needPeerCopy)
 		}
 	}
+
+	// process dimension tables
+	// add dimension tables only if no previously existing shards owned by data node
+	// remove dimension tables only if no shard left for the node
+	if noExistingShards && !noShardLeft {
+		allShardsInitializing := initializingShards > 0 && len(incoming) == initializingShards
+		for _, table := range dimensionTables {
```
Comment (voyager-dw) [inline datanode/datanode.go]: what is the initializing logic here for dimension tables?

File: datanode/bootstrap.go [hunk 2]
```diff
@@ -121,56 +123,20 @@ func (m *bootstrapManagerImpl) Bootstrap() error {
...
-				mutex.Unlock()
-				// unpin table shard after use
-				tableShard.Users.Done()
-			})
-		}
-	}
-	wg.Wait()
-	if !multiErr.Empty() {
-		return multiErr.FinalError()
+	err := m.bootstrapable.Bootstrap(m.peerSource, m.origin, m.topo, topoStateSnapshot, m.opts.BootstrapOptions())
```
Comment (shz117) [inline datanode/bootstrap.go]: is BootstrapManager and Bootstrap 1:1 ?
in that case BootstrapManager seems dummy and can be removed..

File: datanode/datanode.go [hunk 3]
```diff
@@ -257,20 +258,43 @@ func (d *dataNode) startDebugServer() {
...
+	shardOwnershipEvents, done, err := d.metaStore.WatchShardOwnershipEvents()
+	if err != nil {
+		utils.GetLogger().With("error", err.Error()).Fatal("failed to watch schema addition")
+	}
+
+	for {
+		select {
+		case <-d.close:
+			return
+		case event, ok := <-shardOwnershipEvents:
```
Comment (shz117) [inline datanode/datanode.go]: curious what the second return value from channel read is

File: datanode/datanode.go [hunk 4]
```diff
@@ -303,39 +327,52 @@ func (d *dataNode) startAnalyzingShardAvailability() {
...
+				factTables = append(factTables, table)
+			}
+		}
+		d.memStore.RUnlock()
+
 		// check all whether all tables within initialing shards are bootstrapped
 		for shardID := range initializing {
 			numTablesBootstrapped := 0
-			for _, table := range tables {
+			for _, table := range factTables {
```
Comment (shz117) [inline datanode/datanode.go]: who marks dimension table availability?

File: datanode/datanode.go [hunk 5]
```diff
@@ -432,40 +456,90 @@ func (d *dataNode) advertise() {
...
-		d.opts.InstrumentOptions().Logger().Info("start advertising datanode to cluster",
-			zap.String("id", d.hostID))
+		d.logger.Info("start advertising datanode to cluster")
 	}
 }
 
-// GetTableShard returns the table shard from the datanode
-func (d *dataNode) GetTableShard(table string, shardID uint32) (*memstore.TableShard, error) {
-	return d.memStore.GetTableShard(table, int(shardID))
+func (d *dataNode) addTable(table string) {
```
Comment (voyager-dw) [inline datanode/datanode.go]: any conflict or change with the WatchShardOwnershipEvents in recovery?

File: datanode/bootstrap.go (hunk 2 above)
Comment (jshencode) [reply datanode/bootstrap.go]: bootstrapManager still acts as a scheduler

File: datanode/datanode.go (hunk 3 above)
Comment (jshencode) [reply datanode/datanode.go]: when channel is closed, ok will be false

File: datanode/datanode.go (hunk 4 above)
Comment (jshencode) [reply datanode/datanode.go]: under current m3 framework we don't have availability tracking for dimension table (non sharded) out of box. we will need to add dimension table availability tracking ourselves in etcd

File: datanode/datanode.go (hunk 5 above)
Comment (jshencode) [reply datanode/datanode.go]: in distributed mode, we do not call InitShards as we do in single instance mode under MemStore (in recovery.go)

File: datanode/datanode.go (hunk 1 above)
Comment (jshencode) [reply datanode/datanode.go]: dimension tables will be added when first shard is assigned to the data node
and removed when the last shard is removed from the data node

File: datanode/datanode.go (hunk 4 above)
Comment (shz117) [reply datanode/datanode.go]: I see. maybe add a todo somewhere

**Pull request merged**